### ISSUE TITLE
How to modify the loading of clear-degraded image pairs in knowledge distillation

### ISSUE THREAD

**Issue by Wangzhongqiu410** [ultralytics/ultralytics#20478]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I improved yolo and used the idea of ​​offline distillation. I first used clear images to train a teacher model model_t, and then used low-quality images to train the student model model_s. During the training process, the teacher model loaded clear images in eval mode. The student model detects objects on degraded images.
I have now implemented the teacher and student loading the same set of data. I now want to modify the data loading part to load image pairs during training, but I don't know how to modify the dataloader part to meet my needs.
Please tell me which part of the code I should modify, thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Wangzhongqiu410, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖. An Ultralytics engineer will assist you here soon!

**Comment by glenn-jocher** [comment]
Hi @Wangzhongqiu410, 

For implementing offline distillation with paired clear-degraded images, you'll need to create a custom dataset class by extending `BaseDataset`. The main files to focus on are:

1. `ultralytics/data/base.py` - Create a custom dataset class extending `BaseDataset`
2. `ultralytics/data/build.py` - Either modify `build_yolo_dataset` or create your own builder function

In your custom dataset, you'll need to implement a way to:
- Track both clear and degraded image paths (perhaps storing them as parallel lists)
- Override the `load_image` method to load your paired images
- Modify `get_image_and_label` to return both versions

For the trainer, you may need to extend `DetectionTrainer` to use your custom dataset and handle feeding the appropriate images to the teacher and student models during training.

The core modification will be in how you implement the dataset class to manage the paired data structure and provide the correct images to each model.

**Comment by Wangzhongqiu410** [comment]
> 你好[@Wangzhongqiu410](https://github.com/Wangzhongqiu410)，
> 
> 要使用成对的清晰和降级图像实现离线蒸馏，您需要通过扩展来创建自定义数据集类`BaseDataset`。需要关注的主要文件是：
> 
> 1. `ultralytics/data/base.py`- 创建扩展的自定义数据集类`BaseDataset`
> 2. `ultralytics/data/build.py`- 修改`build_yolo_dataset`或创建您自己的构建器函数
> 
> 在您的自定义数据集中，您需要实现一种方法来：
> 
> * 跟踪清晰和退化的图像路径（可能将它们存储为并行列表）
> * 重写`load_image`方法来加载配对的图像
> * 修改`get_image_and_label`以返回两个版本
> 
> 对于训练师，您可能需要扩展`DetectionTrainer`使用自定义数据集，并在训练期间处理向教师和学生模型提供适当的图像。
> 
> 核心修改在于如何实现数据集类来管理配对数据结构并为每个模型提供正确的图像。

Hello, thank you for your reply. I modified BaseDataset and build.py under base.py according to your suggestion and added the loading of sharp_img. However, after I modified it, I found that there was a problem when loading the batch in trainer.py. The obtained batch did not have bbox, keypoints, etc. I don’t know where the problem is. I look forward to your reply.

![Image](https://github.com/user-attachments/assets/1947931e-be5d-4eeb-9476-6e13f89b1213)

**Comment by Wangzhongqiu410** [comment]
Hi, I found the problem here. After I added img-sharp, the bbox, keypoints and other information in the batch obtained in trainer.py `for i, batch in pbar:` are all 0. Why is this? Please give me some suggestions
` if epoch == (self.epochs - self.args.close_mosaic):
                self._close_dataloader_mosaic()
                self.train_loader.reset()

            if RANK in {-1, 0}:
                LOGGER.info(self.progress_string())
                pbar = TQDM(enumerate(self.train_loader), total=nb)
            self.tloss = None
            for i, batch in pbar:
                self.run_callbacks("on_train_batch_start")
                # Warmup
                ni = i + nb * epoch
                if ni <= nw:`
